source/usr/lib/mcvirt/cluster/cluster.py

```diff
@@ -15,437 +17,555 @@
+                node_status = 'Connected'
+            except CouldNotConnectToNodeException:
+                pass
+            table.add_row((node, node_config['ip_address'],
+                           node_status))
+        return table.draw()
+
+    @Pyro4.expose()
+    @locking_method()
+    def add_node_configuration(self, node_name, ip_address,
+                               connection_user, connection_password,
+                               ca_key):
+        """Add MCVirt node to configuration and generates SSH
+        authorized_keys file
+        """
+        self._get_registered_object('auth').assert_permission(PERMISSIONS.MANAGE_CLUSTER)
+
+        # Create CA file
+        ssl_object = self._get_registered_object(
+            'certificate_generator_factory').get_cert_generator(node_name)
+        ssl_object.ca_pub_file = ca_key
+
+        # Connect to node and obtain cluster user
+        remote = Connection(username=connection_user, password=connection_password,
+                            host=node_name)
+        remote_user_factory = remote.get_connection('user_factory')
+        connection_user = remote_user_factory.get_user_by_username(connection_user)
+        remote.annotate_object(connection_user)
+        username, password = connection_user.create_cluster_user(host=get_hostname())
 
-    @staticmethod
-    def getHostname():
-        """Returns the hostname of the system"""
-        return socket.gethostname()
+        # Add node to configuration file
+        def add_node_config(mcvirt_config):
+            mcvirt_config['cluster']['nodes'][node_name] = {
+                'ip_address': ip_address,
+                'username': username,
+                'password': password
+            }
+        MCVirtConfig().update_config(add_node_config)
 
-    def __init__(self, mcvirt_instance):
-        """Sets member variables"""
-        self.mcvirt_instance = mcvirt_instance
+    @Pyro4.expose()
+    @locking_method()
+    def add_node(self, node_connection_string):
+        """Connect to a remote MCVirt machine, shares SSH keys and clusters the machines."""
+        # Ensure the user has privileges to manage the cluster
+        self._get_registered_object('auth').assert_permission(PERMISSIONS.MANAGE_CLUSTER)
 
-        # Connect to each of the nodes
-        if (self.mcvirt_instance.initialise_nodes):
-            self.connectNodes()
+        try:
+            config_json = base64.b64decode(node_connection_string)
+            node_config = json.loads(config_json)
+            assert 'username' in node_config and node_config['username']
+            assert 'password' in node_config and node_config['password']
+            assert 'ip_address' in node_config and node_config['ip_address']
+            assert 'hostname' in node_config and node_config['hostname']
+            assert 'ca_cert' in node_config and node_config['ca_cert']
+        except:
+            raise InvalidConnectionString('Connection string is invalid')
```

Comment:
Could maybe catch specific exceptions rather than all just to be safe? Not sure how important this is though. I think ``base64.b64dencode()`` can raise a ``TypeError``, ``json.loads()`` can raise a ``ValueError`` if not valid JSON, and obviously the asserts could raise an ``AssertionError``.